File: 백준/1번문제/S0okJu/Q11725.py
"""
0. 문제 분석
    i) 루트 없는 트리
    ii) 트리의 루트 = 1
    iii) 입력받을때 순서가 없다.
    iv) 순차적인 인덱스를 가진 노드의 부모 노드를 구해야 한다. 
        -> 리스트의 인덱스를 활용!
    
1. 아이디어 
- 방향성이 없는 트리를 만든 후 하나씩 탐색하여 부모 리스트를 제작한다. 

2. 시간복잡도 :
O(n)
-> 하나 노드 개수만큼 순회하다가 또 순회하고 순회하고를 반복한다. 

3. 자료구조 : Tree

"""

# tree의 1차원 인덱스를 자식 노드로 정하면 되는 것 아닌가?
# 입력 받을때 순서가 랜덤이다. 
# 예외 4 1 -> 표에서는 1이 4보다 Level이 높다. 
def invalid_solution():
    n = int(input())
    tree = [[] for _ in range(n+1)]
    for i in range(n-1):
        a,b = map(int,input().split())
        tree[b].append(a)
    
    for j in range(2,n+1):
        print(tree[j][0])
        
import sys
# 재귀의 최대 깊이
# 없으면 런타임에러(RecursionError)
sys.setrecursionlimit(10**6)

def search(node):
    # 트리의 자식 노드를 모두 탐색 
    for child in tree[node]:
        # 부모 노드 등록 
        if parent_list[child] == 0:
            parent_list[child] = node
            # 자식 노드의 부모 노드를 등록 
            search(child)
        
    
n = int(input())
tree = [[] for _ in range(n+1)]
# 자식 노드의 
parent_list = [0] * (n+1)

# 트리 만들기 
# 방향성이 없으므로 둘 다 정의하자
for i in range(n-1):
    a,b = map(int,input().split())
    tree[b].append(a)
    tree[a].append(b)


search(1)    

for p in range(2,len(parent_list)):
    print(parent_list[p])
    

        

File: 백준/1번문제/최원준/Q11725.java
// https://www.acmicpc.net/problem/
package 백준;

/*

1. 아이디어 :
    그래프를 사용하여 인접한 노드들을 관리합니다.
    루트부터 탐색을 시작하면서 자싟들을 큐에 넣고 정답을 갱신합니다.
2. 시간복잡도 :
    O( n )
3. 자료구조 :
    해시맵, 해시셋

 */
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Q11725 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static String input;
    static String[] inputs;
    static int[] ninputs;

    public static void main(String[] args) throws IOException {
        int n = Integer.parseInt(br.readLine());

        Set<Integer> visited = new HashSet<>();
        HashMap<Integer, Set<Integer>> graph = new HashMap<>();

        for (int i=0; i<n-1; i++) {
            inputs = br.readLine().split(" ");
            int a = Integer.parseInt(inputs[0]), b = Integer.parseInt(inputs[1]);
            graph.putIfAbsent(a, new HashSet<>());
            graph.putIfAbsent(b, new HashSet<>());
            graph.get(a).add(b);
            graph.get(b).add(a);
        }


        int[] ans = new int[n+1];
        Deque<Integer> deque = new ArrayDeque<>();
        deque.addLast(1);
        visited.add(1);

        while (!deque.isEmpty()) {
            int curr = deque.pollFirst();
            for (int connectedNode : graph.get(curr)) {
                if (visited.contains(connectedNode)) continue;
                visited.add(connectedNode);
                deque.addLast(connectedNode);
                ans[connectedNode] = curr;
            }
        }
        for (int i=2; i<=n; i++) {
            System.out.println(ans[i]);
        }
    }
}

/*

import sys
from collections import defaultdict, deque
#sys.setrecursionlimit(1000000)
# input = sys.stdin.readline
input = lambda: sys.stdin.readline().rstrip()

def solution(n, arr):
    graph = defaultdict(set)
    for a, b in arr:
        graph[a].add(b)
        graph[b].add(a)

    ans = [0] * (n+1)
    queue = deque()
    visited = set()
    queue.append(1)
    visited.add(1)

    while queue:
        curr = queue.popleft()
        for child in graph[curr]:
            if child in visited: continue
            visited.add(child)
            queue.append(child)
            ans[child] = curr

    for num in ans[2:]:
        print(num)



if __name__ == '__main__':
    n = int(input())
    arr = []
    for i in range(n-1):
        arr.append(tuple(map(int, input().split())))
    solution(n, arr)
 */